rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1229 @@
+the official Nix formatter must be used.
+
+## Examples and Interactions
+
+### Git blames
+
+Reformatting commits that get added to `.git-blame-ignore-revs` [won't get shown](https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view) in blames on GitHub,
+and can be ignored in the `git blame` command using [`--ignore-revs-file`](https://www.git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt).
+
+### Formatting gotchas requiring manual intervention
+
+Some code patterns in Nixpkgs will result in a sub-optimal format,
+because an auto-formatter cannot do exceptions based on context.
+A lot of the times though, the same program can be equivalently expressed in a prettier way.
+
+#### CLI argument flags
+
+It is common to pass in CLI flags (e.g. to builders) like this:
+
+```nix
+[
+  "--some-flag" "some-value"
+]
+```
+
+However, this will be formatted sub-optimally:
+
+```nix
+[
+  "--some-flag"
+  "some-value"
+]
+```
+
+If the CLI also accepts GNU-style flags, a more structured helper can be used instead:
+
+```nix
+lib.cli.toGNUCommandLine { } {
+  some-flag = "some-value";
+}
+```
+
+#### Singleton lists
+
+Sometimes a list only needs a single element
+and there's no expectation to add more in the future.
+This would be formatted like this:
+
+```nix
+{
+  list = [
+    {
+      foo = 10;
+      bar = 20;
+    }
+  ];
+}
+```
+
+In this case one level of indentation can be saved using [`lib.singleton`](https://nixos.org/manual/nixpkgs/stable#function-library-lib.lists.singleton):
+
+```nix
+{
+  list = lib.singleton {
+    foo = 10;
+    bar = 20;
+  };
+}
+```
+
+## Drawbacks
+
+- No automatic code format can be as pretty as a carefully crafted manual one. There will always be "ugly" edge cases.
+  - However, we argue that on average the new format will be an improvement,
+    and that contributors should not waste their time making the code slightly more pretty.
+- Every formatter will have bugs, but the Nix formatter team will be able to make fixes.
+- Having a commit that changes the formatting, can make git blame harder to use.
+  - This can be [worked around in interfaces](#git-blames).
+
+## Alternatives
+
+- Keep the status quo of not having an official formatter.
+  The danger is that this creates discord within the Nix community.
+  The current friction and maintainer churn due to bad formatting may arguably be small, but not negligible.
+- Specify a different format.
+- Specify a format, but don't enforce it in CI, to allow manually tweaking the output if necessary.
+- Apply the format incrementally, i.e. only changed files get formatted, to make a more graceful transition.
+  However, such an approach would not reduce the amount of merge conflicts,
+  and increase the workload for contributors during that time significantly.
+
+## Unresolved questions
+
+- Should the line length limit be specified?
```

Comment:
I was just experimenting with the current implementation and realized that if I change the line length limit to something small, format my code, then change it back to default, some of the now multi-line expression get brought back to one-liners. If this is going to be the final behavior, I don't think line length should be configurable, as it no longer creates a standard, but multiple incompatible standards. If however, formats from smaller line length limits are unchanged by larger line length limits, I think keeping it configurable is probably okay.

Reply:
I suppose another solution could be to keep the current behavior, and make it configurable on a per file basis via a comment in the file, but I can't say I'm a huge fan of that.

Reply:
I won't go into whether or not the formatter should do this, as it has been extensively discussed in https://github.com/NixOS/rfcs/pull/166#discussion_r1394844307 already.

About the question of the implications for the configurable line length limit however:

I think that it is fine as-is. The format detailed in the RFC only focuses on the format of the expanded (multi-line) form of expressions, together with "Unless stated otherwise, any expression that fits onto one single line will be trivially formatted as such".

This deliberately leaves a lot of freeroom to define what means to "fit onto one single line" in many but not all cases, and thus allows to configure the line length limit. I'd disagree with your "multiple incompatible standards",  and instead of think of it as different "flavors" of the same standard.

Reply:
The reason I do not see it as different flavors of the same standard is this: nixpkgs is going to have to pick a line length setting. If nixpkgs is going to be linted based on this standard, developers will not be able to submit code that is formatted with any other "flavor". And If a particular formatting doesn't qualify for use in nixpkgs, I have a hard time seeing it as "standard".

Reply:
Yes, there will be one "blessed" configuration used for Nixpkgs itself, and it will be pinned in the repo with tooling for everyone to automatically use it. This doesn't mean other Nix related projects shouldn't be able to set a different line length limit if desired.

Reply:
Furthermore, third-party projects probably don't want to use a vastly different line length limit than the default one, so submitting to Nixpkgs will require few changes, while the overall style and feeling of the format is the same.

We think this is fine and not a big problem overall.